pkg-manager/plugin-commands-installation/src/dedupeDiffCheck.ts

```diff
@@ -0,0 +1,107 @@
+  }
+
+  const changesCount =
+    countChangedSnapshots(issues.importerIssuesByImporterId) +
+    countChangedSnapshots(issues.packageIssuesByDepPath)
+
+  if (changesCount > 0) {
+    throw new DedupeCheckIssuesError(issues)
+  }
+}
+
+export class DedupeCheckIssuesError extends PnpmError {
+  constructor (public dedupeCheckIssues: DedupeCheckIssues) {
+    super('DEDUPE_CHECK_ISSUES', 'Dedupe --check found changes to the lockfile')
+  }
+}
+
+/**
+ * Get all the keys of an object T where the value extends some type U.
+ */
+type KeysOfValue<T, U> = KeyValueMatch<T, keyof T, U>
+type KeyValueMatch<T, K, U> = K extends keyof T
+  ? T[K] extends U ? K : never
+  : never
+
+/**
+ * Given a PackageSnapshot or ProjectSnapshot, returns the keys where values
+ * match ResolvedDependencies.
+ *
+ * Unfortunately the ResolvedDependencies interface is just
+ * Record<string,string> so this also matches the "engines" and "specifiers"
+ * block.
+ */
+type PossiblyResolvedDependenciesKeys<TSnapshot> = KeysOfValue<TSnapshot, ResolvedDependencies | undefined>
+
+function diffSnapshots<TSnapshot> (
+  prev: Record<string, TSnapshot>,
+  next: Record<string, TSnapshot>,
+  fields: ReadonlyArray<PossiblyResolvedDependenciesKeys<TSnapshot>>
+): SnapshotsChanges {
+  const removed: string[] = []
+  const updated: Record<string, ResolutionChangesByAlias> = {}
+
+  for (const [id, prevSnapshot] of Object.entries(prev)) {
+    const nextSnapshot = next[id]
+
+    if (nextSnapshot == null) {
+      removed.push(id)
+      continue
+    }
+
+    const updates = fields.reduce((acc: ResolutionChangesByAlias, dependencyField) => ({
+      ...acc,
+      ...getResolutionUpdates(prevSnapshot[dependencyField] ?? {}, nextSnapshot[dependencyField] ?? {}),
+    }), {})
+
+    if (Object.keys(updates).length > 0) {
+      updated[id] = updates
+    }
+  }
+
+  const added = Object.keys(next).filter(id => prev[id] == null)
+
+  return { added, removed, updated }
+}
+
+function getResolutionUpdates (prev: ResolvedDependencies, next: ResolvedDependencies): ResolutionChangesByAlias {
+  const updates: ResolutionChangesByAlias = {}
+
+  for (const [alias, prevResolution] of Object.entries(prev)) {
+    const nextResolution = next[alias]
+
+    if (prevResolution === nextResolution) {
+      continue
+    }
+
+    updates[alias] = nextResolution == null
+      ? { type: 'removed', prev: prevResolution }
+      : { type: 'updated', prev: prevResolution, next: nextResolution }
+  }
+
+  const newAliases = Object.entries(next).filter(([alias]) => prev[alias] == null)
+  for (const [alias, nextResolution] of newAliases) {
+    updates[alias] = { type: 'added', next: nextResolution }
+  }
+
+  return updates
+}
+
+function countChangedSnapshots (snapshotChanges: SnapshotsChanges): number {
+  return snapshotChanges.added.length + snapshotChanges.removed.length + Object.keys(snapshotChanges.updated).length
+}
```

Comment:
isn't the same called in the renderer? Maybe the total count should be a field in SnapshotsChanges?

Reply:
Yup it is the same call. I wasn't sure if it'd be better to add a dependency or copy a short function.

Decided to remove this from `@pnpm/dedupe.issues-renderer` in the latest push and check for whether `importersReport` or `packagesReport` are empty strings instead. I think that's okay since the reporting is mostly operating off of strings anyway.

> Maybe the total count should be a field in SnapshotsChanges?

I was a bit worried about storing derived data. It's a common programming mistake for code written in the future to not remember to update fields in derived data when changing other fields.